api/models/scanstatus.go

```diff
@@ -0,0 +1,115 @@
+package models
+
+import (
+	"fmt"
+	"time"
+)
+
+var scanStatusStateTransitions = map[ScanStatusState][]ScanStatusState{
+	ScanStatusStatePending: {
+		ScanStatusStateDiscovered,
+		ScanStatusStateAborted,
+		ScanStatusStateFailed,
+		ScanStatusStateDone,
+	},
+	ScanStatusStateDiscovered: {
+		ScanStatusStateInProgress,
+		ScanStatusStateAborted,
+	},
+	ScanStatusStateInProgress: {
+		ScanStatusStateAborted,
+		ScanStatusStateFailed,
+		ScanStatusStateDone,
+	},
+	ScanStatusStateAborted: {
+		ScanStatusStateFailed,
+	},
+}
+
+var scanStatusReasonMapping = map[ScanStatusState][]ScanStatusReason{
+	ScanStatusStatePending: {
+		ScanStatusReasonCreated,
+	},
+	ScanStatusStateDiscovered: {
+		ScanStatusReasonAssetsDiscovered,
+	},
+	ScanStatusStateInProgress: {
+		ScanStatusReasonAssetScansRunning,
+	},
+	ScanStatusStateAborted: {
+		ScanStatusReasonCancellation,
+	},
+	ScanStatusStateFailed: {
+		ScanStatusReasonCancellation,
+		ScanStatusReasonAssetScanFailed,
+		ScanStatusReasonTimeout,
+	},
+	ScanStatusStateDone: {
+		ScanStatusReasonNothingToScan,
+		ScanStatusReasonSuccess,
+	},
+}
+
+func NewScanStatus(s ScanStatusState, r ScanStatusReason, m *string) *ScanStatus {
+	return &ScanStatus{
+		State:              s,
+		Reason:             r,
+		Message:            m,
+		LastTransitionTime: time.Now(),
+	}
+}
+
+func (a *ScanStatus) Equals(aa *ScanStatus) bool {
+	if a.Message == nil && aa.Message != nil {
+		return false
+	}
+	if aa.Message == nil && a.Message != nil {
+		return false
+	}
+	if a.Message == nil && aa.Message == nil {
+		return a.State == aa.State && a.Reason == aa.Reason
+	}
+
+	return a.State == aa.State && a.Reason == aa.Reason && *a.Message == *aa.Message
+}
+
+func (a *ScanStatus) isValidStatusTransition(aa *ScanStatus) error {
+	transitions, ok := scanStatusStateTransitions[a.State]
+	if ok {
+		for _, transition := range transitions {
+			if transition == aa.State {
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("invalid transition: from=%s to=%s", a.State, aa.State)
+}
+
+func (a *ScanStatus) isValidReason(aa *ScanStatus) error {
+	reasons, ok := scanStatusReasonMapping[aa.State]
+	if ok {
+		for _, reason := range reasons {
+			if reason == aa.Reason {
+				return nil
+			}
+		}
+	}
```

Comment:
this can serve as a validator since it does not require another object for checking the reason
```suggestion
func (a *ScanStatus) isValid() error {
    // check if the reason is allowed for the state
	reasons, _ := scanStatusReasonMapping[a.State]
	for _, reason := range reasons {
		if reason == a.Reason {
			return nil
		}
	}

	return fmt.Errorf("invalid reason for state: state=%s reason=%s", a.State, a.Reason)
}
```

Reply:
Not sure how this function can replace the current one, can you elaborate on the caller side of your version please?

Reply:
Sure, added an example for this in the new review